common/web/gesture-recognizer/src/gestureRecognizer.ts

```diff
@@ -0,0 +1,64 @@
+/// <reference path="gestureRecognizerConfiguration.ts" />
+/// <reference path="includes/events.ts" />
+
+namespace com.keyman.osk {
+  type Mutable<Type> = {
+    -readonly [Property in keyof Type]: Type[Property];
+  };
+
+  export enum TrackedInputState {
+    START = "start",
+    END = "end",
+    MOVE = "move",
+    CANCEL = "cancel"
+  };
+
+  export class GestureRecognizer extends EventEmitter {
+    public static readonly TRACKED_INPUT_UPDATE_EVENT_NAME = "trackedInputUpdate";
+
+    protected readonly config: GestureRecognizerConfiguration;
+
+    private readonly mouseEngine?: MouseEventEngine;
+    private readonly touchEngine?: TouchEventEngine;
+
+    protected static preprocessConfig(config: GestureRecognizerConfiguration): GestureRecognizerConfiguration {
+      // Allows configuration pre-processing during this method.
+      let processingConfig: Mutable<GestureRecognizerConfiguration> = Object.assign({}, config);
+      processingConfig.mouseEventRoot = processingConfig.mouseEventRoot ?? processingConfig.targetRoot;
+      processingConfig.touchEventRoot = processingConfig.touchEventRoot ?? processingConfig.targetRoot;
+
+      return processingConfig;
+    }
+
+    public constructor(config: GestureRecognizerConfiguration) {
+      super();
+      config = GestureRecognizer.preprocessConfig(config);
+      this.config = config;
+
+      this.mouseEngine = new MouseEventEngine(config);
+      this.touchEngine = new TouchEventEngine(config);
+
+      this.mouseEngine?.registerEventHandlers();
+      this.touchEngine?.registerEventHandlers();
+
+      const _this = this;
+      this.mouseEngine?.on(InputEventEngine.INPUT_UPDATE_EVENT_NAME, function(state, coord) {
+        _this.emit(GestureRecognizer.TRACKED_INPUT_UPDATE_EVENT_NAME, state, coord);
+        return false;
+      });
+
+      this.touchEngine?.on(InputEventEngine.INPUT_UPDATE_EVENT_NAME, function(state, coord) {
+        _this.emit(GestureRecognizer.TRACKED_INPUT_UPDATE_EVENT_NAME, state, coord);
+        return false;
+      });
+    }
+
+    public destroy() {
+      this.mouseEngine?.unregisterEventHandlers();
+      this.touchEngine?.unregisterEventHandlers();
+
+      (this.mouseEngine as any) = null;
+      (this.touchEngine as any) = null;
```

Comment:
I don't get why this casting is needed... It looks weird.

Reply:
```
private readonly mouseEngine:  MouseEventEngine;
```

As it's marked readonly, we're not allowed to modify it directly without odd shenanigans like this.  Though, maybe that `Mutable` pattern could do the trick...

At any rate, I'm bypassing `readonly` for the `destroy` method, as it's intended to fill a destructor-type role to help prevent potential memory leaks, especially during unit testing.

Reply:
... and nope, `Mutable` won't help here.  It relies on `keyof` operations, which TS actively prevents from accessing protected and private fields - with good reason, admittedly.

Reply:
OK. Can you add a code comment on the reason for the cast because I don't think it will be obvious to future maintainers. :)